subsys/zigbee/cli/zigbee_cli_cmd_zcl_groups.c

```diff
@@ -0,0 +1,996 @@
+	ZB_ERROR_CHECK(zb_err_code);
+
+	ctx_mgr_delete_entry(entry);
+exit:
+	zb_buf_free(bufid);
+}
+
+/**@brief Function to check if command is to be sent as unicast.
+ *
+ * @param[in] packet_info   Pointer to structure with packet data.
+ *
+ * @return ZB_TRUE if command is sent as unicast, ZB_FALSE otherwise.
+ */
+static zb_bool_t is_sent_unicast(struct zcl_packet_info *packet_info)
+{
+	switch (packet_info->dst_addr_mode) {
+	case ADDR_LONG:
+		return ZB_TRUE;
+	case ADDR_SHORT:
+		if (ZB_NWK_IS_ADDRESS_BROADCAST(packet_info->dst_addr.addr_short)) {
+			return ZB_FALSE;
+		} else if ((packet_info->dst_ep < ZB_MIN_ENDPOINT_NUMBER) ||
+		    (packet_info->dst_ep == ZB_APS_BROADCAST_ENDPOINT_NUMBER)) {
+			return ZB_FALSE;
+		} else {
+			return ZB_TRUE;
+		}
+	default:
+		return ZB_FALSE;
+	}
+}
+
+/**@brief Add remote node to the group or remove from the group.
+ *
+ * @code
+ * zcl group add <h:dst_addr> <d:ep> [-p h:profile] <h:group_id>
+ * @endcode
+ *
+ * Send Add Group command to the endpoint `ep` of the remote node `dst_addr`
+ * and add it to the group `group_id`.
+ * Providing profile ID is optional, Home Automation Profile is used by default.
+ *
+ * @code
+ * zcl group remove <h:dst_addr> <d:ep> [-p h:profile] <h:group_id>
+ * @endcode
+ *
+ * Send Remove Group command to the endpoint `ep` of the remote node `dst_addr`
+ * and remove it from the group `group_id`.
+ * Providing profile ID is optional, Home Automation Profile is used by default.
+ *
+ * @note These commands can be sent as groupcast. To send command as groupcast,
+ *       set `dst_addr` to group address and `ep` to 0.
+ *
+ * @note Group name is not supported.
+ */
+int cmd_zb_add_remove_group(const struct shell *shell, size_t argc, char **argv)
+{
+	zb_bufid_t bufid;
+	struct groups_cmd_data *req_data;
+	struct zcl_packet_info *packet_info;
+	struct ctx_entry *entry = ctx_mgr_new_entry(CTX_MGR_GROUPS_CMD_ENTRY_TYPE);
+
+	if (entry == NULL) {
+		zb_cli_print_error(shell, "Request pool empty - wait a bit", ZB_FALSE);
+		return -ENOEXEC;
+	}
+
+	req_data = &(entry->groups_cmd_data);
+	packet_info = &(req_data->packet_info);
+
+	/*  Set default profile ID to Home Automation. */
+	packet_info->prof_id = ZB_AF_HA_PROFILE_ID;
+	/* Set function to be used to send Add Group or Remove Group command. */
+	if (strcmp(*(argv), "add") == 0) {
+		req_data->send_fn = zb_zcl_send_add_group_cmd;
+	} else if (strcmp(*(argv), "remove") == 0) {
+		req_data->send_fn = zb_zcl_send_remove_group_cmd;
+	} else {
+		zb_cli_print_error(entry->shell, "Failed to parse command name", ZB_FALSE);
+		goto add_remove_group_error;
+	}
+
+	packet_info->dst_addr_mode = parse_address(*(++argv), &(packet_info->dst_addr), ADDR_ANY);
+
+	if (packet_info->dst_addr_mode == ADDR_INVALID) {
+		zb_cli_print_error(shell, "Invalid address", ZB_FALSE);
+		goto add_remove_group_error;
+	}
+
+	if (!zb_cli_sscan_uint8(*(++argv), &(packet_info->dst_ep))) {
```

Comment:
Please improve data validatiion. dst_ep is not validated properly so it's possible to pass data like -1 or FF which ends in ZEPHYR FATAL ERROR